src/controllers/sendResetToken.js

```diff
@@ -0,0 +1,109 @@
+import { randomBytes } from 'crypto';
+import user from '../db/models';
+import sendgrid from '../middlewares/helper';
+
+const { User } = user;
+
+/**
+  * @description class representing reset user password
+  *
+  * @class SendResettoken
+  */
+class SendResetToken {
+  /**
+    * @description - This method is responsible for creating creating password reset token
+    *
+    * @static
+    * @param {object} request - Request sent to the router
+    * @param {object} response - Response sent from the controller
+    *
+    * @returns {object} - object representing response message
+    *
+    * @memberof UserHandler
+    */
+  static async resetToken(request, response) {
+    try {
+      const { email } = request.body;
+      const userData = await User.findOne({
+        where: { email }
+      });
+
+      if (!userData) {
+        return response.status(400).json({
+          status: 'Fail',
+          message: 'An email has been sent to your mail',
+          token: request.body.token
+        });
+      }
+      const token = randomBytes(40).toString('hex');
+      const tokenExpiresIn = Date.now() + 3600000;
+
+      const { username, password } = userData.dataValues;
+      const userUpdatedData = await userData.update({
+        token,
+        username,
+        email,
+        password,
+        passwordExpiresIn: tokenExpiresIn || userData.token,
+      });
+      const { HOST_URL } = process.env;
+      await sendgrid(userData.email, 'noreply@authors.com', HOST_URL, token);
+      response.status(200).json({
+        status: 'Success',
+        message: 'Password reset link has been sent to your mail',
+        token: userUpdatedData.token,
+      });
+    } catch (error) {
+      response.status(500).json({
+        status: 'Error',
+        error: error.message
+      });
+    }
+  }
+
+  /**
+    * @description - This method is responsible for updating user password
+    *
+    * @static
+    * @param {object} request - Request sent to the router
+    * @param {object} response - Response sent from the controller
+    *
+    * @returns {object} - object representing response message
+    *
+    * @memberof UserHandler
+    */
+  static async updateUserPassword(request, response) {
+    const { password, confirmNewPassword, token } = request.body;
+    try {
+      const userToUpdate = await User.findOne({
+        where: { token }
+      });
+      if (parseInt(userToUpdate.passwordExpiresIn, 10) < parseInt(Date.now(), 10)) {
+        return response.status(400).json({
+          status: 'Fail',
+          message: 'The link has expired',
+        });
+      }
+      if (password !== confirmNewPassword) {
+        return response.status(400).json({
+          status: 'Fail',
+          message: 'The passwords do not match, confirm and type again.'
+        });
+      }
+      await userToUpdate.update({
+        password: password || userToUpdate.password,
```

Comment:
Nice job setting a fallback password here, but if you think about it, `userToUpdate.password` returns the hashed password that was saved to the database when the user first joined.

Now, per the `beforeUpdate` hook you set on the `User` model where passwords are being hashed right before they are updated, isn't that going to be hashing an already hashed password? Don't you think that is problematic?

Reply:
resolved.